pyg_lib/csrc/sampler/subgraph.cpp

```diff
@@ -25,10 +28,76 @@ std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>> subgraph(
   return op.call(rowptr, col, nodes, return_edge_id);
 }
 
+std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>>
+subgraph_bipartite(const at::Tensor& rowptr,
+                   const at::Tensor& col,
+                   const at::Tensor& src_nodes,
+                   const at::Tensor& dst_nodes,
+                   const bool return_edge_id) {
+  at::TensorArg rowptr_t{rowptr, "rowptr", 1};
+  at::TensorArg col_t{col, "col", 1};
+  at::TensorArg src_nodes_t{src_nodes, "src_nodes", 1};
+  at::TensorArg dst_nodes_t{dst_nodes, "dst_nodes", 1};
+
+  at::CheckedFrom c = "subgraph_bipartite";
+  at::checkAllDefined(c, {rowptr_t, col_t, src_nodes_t, dst_nodes_t});
+  at::checkAllSameType(c, {rowptr_t, col_t, src_nodes_t, dst_nodes_t});
+
+  static auto op = c10::Dispatcher::singleton()
+                       .findSchemaOrThrow("pyg::subgraph_bipartite", "")
+                       .typed<decltype(subgraph_bipartite)>();
+  return op.call(rowptr, col, src_nodes, dst_nodes, return_edge_id);
+}
+
+c10::Dict<utils::EdgeType,
+          std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>>>
+hetero_subgraph(const utils::EdgeTensorDict& rowptr,
+                const utils::EdgeTensorDict& col,
+                const utils::NodeTensorDict& src_nodes,
```

Comment:
Not sure why we have both `src_nodes` and `dst_nodes`. IMO, these can be safely merged as in https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.HeteroData.subgraph.

Reply:
Separating `src` and `dst` is just to give some flexibility. We could also have the merged API though.